internal/tpm/session.go

```diff
@@ -0,0 +1,823 @@
+}
+
+func (s *tpmSession) Clear() error {
+	// only the lockout and platform hierarchies can invoke tpm2.Clear
+	// it is possible to block the lockout hierarchy from performing a clear operation
+	// it is possible to add passwords to both the lockout and platform hierarchies
+	// We make a best effort to invoke clear, but if those operations fail, we attempt
+	// to reset owner password and make our keys unrecoverable.
+	hierarchies := []struct {
+		name   string
+		handle tpm2.TPMHandle
+	}{
+		{
+			name:   "lockout",
+			handle: tpm2.TPMRHLockout,
+		},
+		{
+			name:   "platform",
+			handle: tpm2.TPMRHPlatform,
+		},
+	}
+
+	var errs []error
+	for _, hier := range hierarchies {
+		cmd := tpm2.Clear{
+			AuthHandle: tpm2.AuthHandle{
+				Handle: hier.handle,
+				Auth:   tpm2.PasswordAuth(nil),
+			},
+		}
+
+		_, err := cmd.Execute(transport.FromReadWriter(s.conn))
+		if err != nil {
+			errs = append(errs, fmt.Errorf("clearing hierarchy %q: %w", hier.name, err))
+		}
+	}
+
+	// if all commands failed we failed to invoke clear
+	// so we try to clean up as much as we manually can
+	// but still treat everything as if it errored.
+	if len(errs) == len(hierarchies) {
+		// reset the error into a compact one
+		errs = []error{
+			fmt.Errorf("clearing hierarchy failed: %w", errors.Join(errs...)),
+		}
+		if err := s.resetStorageHierarchyPassword(); err != nil {
+			// if we fail to reset the storage password something is very off. We shouldn't erase our
+			// password in case we can try again.
+			return fmt.Errorf("resetting storage hierarchy password: %w %w", err, errors.Join(errs...))
+		}
+		if flushErrs := s.flushKeys(); len(flushErrs) != 0 {
+			errs = append(errs, fmt.Errorf("flushing hierarchy keys: %w", errors.Join(flushErrs...)))
+		}
+	} else {
+		// if any of the above commands succeeded we consider the operation successful
+		errs = nil
+	}
+
+	// clear internal state
+	s.handles = make(map[KeyType]*tpm2.NamedHandle)
+	s.srk = nil
+
+	// clear persistent storage
+	if s.storage != nil {
+		if err := s.storage.ClearPassword(); err != nil {
+			errs = append(errs, fmt.Errorf("clearing stored password: %w", err))
+		}
+		// Clear stored keys by storing empty values
+		if err := s.clearStoredKeys(); err != nil {
+			errs = append(errs, fmt.Errorf("clearing stored keys: %w", err))
+		}
+	}
+
+	if len(errs) != 0 {
+		return errors.Join(errs...)
+	}
+	return nil
+}
+
+func (s *tpmSession) resetStorageHierarchyPassword() error {
+	// Generate a new random password to make keys unrecoverable
+	newPassword, err := s.generateStoragePassword()
+	if err != nil {
+		return fmt.Errorf("generating new password: %w", err)
+	}
+
+	// Get current password from storage
+	currentPassword, err := s.storage.GetPassword()
+	if err != nil {
+		// If no password stored, assume empty password
+		currentPassword = []byte{}
+	}
+
+	// Update the hierarchy password
+	if err := s.updateStorageHierarchyPassword(currentPassword, newPassword); err != nil {
```

Comment:
Is the intention to leave things in a state that no one can use? If we change the password to a new random value, the device will be left in a state that requires someone to manually clear the tpm in order for it to become usable again. 

Also, this should probably respect EnableOwnership so that we don't set a random password when we didn't have one in the first place.

Reply:
honestly auth is not polished as you can see so stabilization there will come as soon as I get non auth chain validation.

Reply:
the non auth flow is fairly stable will try to realign now